src/parameterizations/vertical/MOM_energetic_PBL.F90

```diff
@@ -2673,6 +2703,233 @@ subroutine ePBL_BBL_column(h, dz, u, v, T0, S0, dSV_dT, dSV_dS, SpV_dt, absf, &
+  real, intent(in) :: u_star    !< The surface friction velocity [Z T-1 ~> m s-1].
+  real, intent(in) :: B_Flux    !< The surface buoyancy flux [Z2 T-3 ~> m2 s-3]
+  real, dimension(SZK_(GV)), intent(in)  :: dz     !< The vertical distance across layers [Z ~> m].
+  real, intent(in) :: MLD_guess !< Mixing Layer depth guessed/found for iteration [Z ~> m].
+  real, dimension(SZK_(GV)+1) :: hz !< depth variable, only used in this routine [H ~> m]
+
+  ! local variables for this subroutine
+  integer :: nz
+  integer :: K ! integer for looping
+  integer :: i,j,n ! integer for looping, local only
+  real :: p1 ! ((B_flux * h))/(u_star^3), boundary layer depth by M-O depth, [nondim]
+  real :: p2 ! ((h f)/u_star ),  boundary layer depth by Ekman depth, [nondim]
+  real :: sm ! sigma_max: location of maximum of shape function in sigma coordinate [nondim]
+  real :: sm_I ! inverse of sm,[nondim]
+  real :: sm_I2 ! An inverse variable given by 1.0/(1.0 - sm), [nondim]
+  real :: hbl ! Boundary layer depth, same as MLD_guess [Z ~> m]
+  real :: F ! function, used in asymptotic model for sm, Equation 7 in Sane et al. 2024 [nondim]
+  real :: F_I ! Inverse of F, [nondim]
+  real :: Fp1 ! = F*p1, [nondim]
+
+  nz = SZK_(GV)+1
+  hz(1) = 0.0
+  do K=2,nz
+    hz(K) = hz(K-1) + dz(K-1)
+  end do
+  hbl = MLD_Guess ! hbl is boundary layer depth.
+  shape_func(:) = 0.0  ! initializing the entire shape_function array
+
+  p1 = (hbl * absf)/(u_star)   ! Boundary layer depth divided by Ekman depth
+  p2 = ((-B_flux * hbl))/(u_star**3) ! Boundary layer depth divided by Monin-Obukhov depth
+  ! B_flux given negative sign to follow convention used in Sane et al. 2023,2024
+  ! p2 < 0 --> surface stabilizing i.e. heating, and p2 > 0 --> surface destabilizing i.e. cooling
+  p2 = p2 * 2.4390 ! dividing by von-Karman constant 0.41 i.e. multiply by 2.4390
+  ! This capping does not matter because these equations have asymptotes. Not sensitive beyond the caps.
+  p1 = min(p1, 2.0) ! capping p1 to less than 2.0. It is always >0.0.
+  p2 = max(p2, -8.0) ! capping p2 to -8.0 if less than -8.0
+  p2 = min(p2,  8.0) ! capping p2 to 8.0 if greater than 8.0
+  ! Empirical model to predict sm:
+  ! F1 is solely function of p2
+  F = exp( (-p2-CS%ML_c(6))/ CS%ML_c(7) ) ! originally, F=(CS%c4/(CS%c5+exp((-p2-CS%c6)/CS%c7)))+CS%c8
+  F = CS%ML_c(5) + F
+  F = CS%ML_c(4) / F
+  F = F + CS%ML_c(8)
+  Fp1 = F*p1
+  Fp1 = max(Fp1, 1E-05) ! an arbitrary small value to cap Fp1, result insensitive below that value
+  F_I = 1.0 / ( Fp1 )
+  sm = CS%ML_c(2) + (CS%ML_c(3) * F_I)
+  sm = CS%ML_c(1) / sm
+  sm = min(sm,0.7) ! makes sure sm is less than 0.7, true sm range is from 0.2 to 0.60
+  sm = max(sm,0.1) ! makes sure sm is more than 0.1
+  sm= sm * hbl ! peak of shape function in model vertical coordinate z, or peak of shape function in z coordinate
+  sm_I = 1.0/sm ! inverse of sm x hbl
+  sm_I2 = 1.0/(hbl-sm)  ! inverse of (hbl-sm), as 0.1<sm<0.7, hbl>sm, hence (hbl-sm) always >0.0
+  
```

Comment:
The entire block of code from lines 2740 to 2764 should be refactored to use far fewer divisions (I suspect that only 3 are needed) with mathematically equivalent expressions.  Doing so would be much more efficient, clearer to read, and it would likely give more robust expressions that do not require as many hard-coded constant limits (e.g., the 1E-05 could probably go).  Come talk to me if you want help doing this.